Adds a SynchronizationMechanism interface for syncing variable values

miguelandres opened the pull request:
- A single, very simple, local-only non-persisting implementation of SynchronizationMechanism was added as LocalValueSyncing and set as the default.
  - Takes over the responsibility of making sure that RemixerItems being added to remixer are compatible with what is already there from the Remixer class.
  - Takes over responsibility of making sure that variables being added to remixer have the latest value (and therefore, remixer can stop keeping track of the latest known value).
- These changes simplify Remixer and other remixer_core classes significantly and reduce the amount of implicit contracts:
  - Remixer no longer keeps dead variables around just to keep their value or configuration.
  - There is never any need to clear callbacks on RemixerItems, because the RemixerItem can be deleted as soon as its parent activity is destroyed.
  - The Remixer.addItem method's logic is simplified.
  - Compatibility checks have been moved to the serializable version of variables and items (StoredVariable) where it's far simpler.
- All test methods testing variable synchronization have been moved to remixer_storage and they are testing the interaction between Remixer and the SynchronizationMechanism, once there are more implementations of these interfaces these tests should be copied for each of them.
- RemixerCallbacks has been renamed to RemixerActivityLifecycleCallbacks so it isn't confused with com.google.android.libraries.remixer.Callback. It was a bad name to begin with.

Commit:
Adds a SynchronizationMechanism interface that is responsible for keeping variables in sync.

- A single, very simple, local-only non-persisting implementation of SynchronizationMechanism was added as LocalValueSyncing in remixer_storage.
  - Moves the responsibility of making sure that RemixerItems being added to remixer are compatible with what is already there from the Remixer class into implementations of SynchronizationMechanism.
  - Moves the responsibility of making sure that variables being added to remixer have the latest value (and therefore, keeping track of the latest known value) to the SynchronizationMechanism implementation.
- These changes simplify Remixer and other remixer_core classes significatively and reduce the amount of implicit contracts:
  - Remixer no longer keeps dead variables around just to keep their value or configuration.
  - There is never any need to clear callbacks on RemixerItems, because the RemixerItem can be deleted as soon as its parent activity is destroyed.
  - The Remixer.addItem method's logic is simplified.
  - Compatibility checks have been moved to the serializable version of variables and items (StoredVariable) where it's far simpler.
- All test methods testing variable synchronization have been moved to remixer_storage and they are testing the interaction between Remixer and the SynchronizationMechanism, once there are more implementations of these interfaces these tests should be copied for each of them.
- RemixerCallbacks has been renamed to RemixerActivityLifecycleCallbacks so it isn't confused with com.google.android.libraries.remixer.Callback. It was a bad name to begin with.

pingpongboss commented on remixer_core/src/main/java/com/google/android/libraries/remixer/Remixer.java:
```diff
@@ -53,12 +60,27 @@ public static Remixer getInstance() {
+   */
+  public void setSynchronizationMechanism(SynchronizationMechanism synchronizationMechanism) {
+    this.synchronizationMechanism = synchronizationMechanism;
+    synchronizationMechanism.setRemixerInstance(this);
```
since this is public API, you may want to guard/handle the case where a client calls this method when there is already a non-null synchronizationMechanism.

Guard:

```
if (this.synchronizationMechanism != null) {
  throw new IllegalStateException();
}
```

Handle:

```
if (this.synchronizationMechanism != null) {
  this.synchronizationMechanism.setRemixerInstance(null); 
}
this.synchronizationMechanism = synchronizationMechanism;
synchronizationMechanism.setRemixerInstance(this);
```

miguelandres replied:
Guarding

pingpongboss commented on remixer_core/src/main/java/com/google/android/libraries/remixer/Remixer.java:
```diff
@@ -77,47 +99,28 @@ public static Remixer getInstance() {
-      // Make newVariable have the current value found in variables that have already been added to
-      // the Remixer.
-      newVariable.setValueWithoutNotifyingOthers(otherVariable.getSelectedValue());
+    if (synchronizationMechanism != null) {
```
synchronizationMechanism == null seems like an error condition to me. What would happen if the client forgets to set a synchronizationMechanism?

miguelandres replied:
Nothing terrible, but data would not be saved/synced and the values across different contexts will be completely independent. You could even use the same keys for different data types in different activities, but nothing would break other than actual syncing or persistence.

In fact, without a synchronization mechanism it would allow for the simplest use of Remixer (I just wanna remix something but not care about keys, or persistence, or many rules). I don't want to encourage that use but not necessarily penalize it either.

pingpongboss commented on remixer_core/src/main/java/com/google/android/libraries/remixer/Remixer.java:
```diff
@@ -53,12 +60,27 @@ public static Remixer getInstance() {
+   * external sources.
+   *
+   * <p>Remixer relies on a SynchronizationMechanism instance to be the source of truth of the
+   * values and configuration, so the user should always set a SynchronizationMechanism.
```
it's not complete clear from this javadoc, but does the user really have to manually set a synchronizationMechanism? I can totally see how users would forget to do this.

It also bothers me that you have this public setter on Remixer which is a singleton. What happens if multiple parts of the user's app calls setSynchronizationMechanism()?

miguelandres replied:
I'll improve the documentation.

pingpongboss replied:
"so the user should always set a SynchronizationMechanism" no longer applies

pingpongboss replied:
You also don't want to throw the exception anymore

pingpongboss commented on remixer_core/src/main/java/com/google/android/libraries/remixer/Remixer.java:
```diff
@@ -77,47 +99,28 @@ public static Remixer getInstance() {
+                "Duplicate key %s being used in class %s",
+                remixerItem.getKey(),
+                remixerItem.getParentObject().getClass().getCanonicalName()
+            ));
```
could this logic also belong in synchronizationMechanism? Could listForKey belong in synchronizationMechanism?

synchronizationMechanism is supposed to "Takes over the responsibility of making sure that RemixerItems being added to remixer are compatible with what is already there from the Remixer class."

miguelandres replied:
Synchronization mechanisms are meant to deal only with the data, not the callbacks and runtime, that's what the core is.

miguelandres replied:
Maybe remixer_core should be remixer_runtime and remixer_storage should be remixer_sync, if that makes more sense?

pingpongboss replied:
remixer_core and remixer_sync make a lot of sense to me.

pingpongboss commented on remixer_core/src/main/java/com/google/android/libraries/remixer/Remixer.java:
```diff
@@ -132,6 +135,34 @@ public void addItem(RemixerItem remixerItem) {
+    List<RemixerItem> itemList = getItemsWithKey(variable.getKey());
+    for (RemixerItem item : itemList) {
+      if (item != variable) {
+        ((Variable) item).setValueWithoutNotifyingOthers(variable.getSelectedValue());
```
seems like Remixer is still responsible for triggering other variables when one changes. Any way the synchronizationMechanism can just take care of this?

miguelandres replied:
It could, I wanted to avoid it though, since that means we need to make sure every implementation of SynchronizationMechanism has to do it and we may need to copy code around.

However that may be the more correct way to deal with this and a proper inheritance tree between them may aid with this. It also ties the SynchronizationMechanism to private methods of Variable and Trigger, I was thinking of a way to avoid using setValueWithoutNotifyingOthers at all from the SynchronizationMechanism

What do you think? do you prefer to move this to the SynchronizationMechanism or to keep it here and then make those internal methods package-local to close access as much as possible?

I actually find it hard to decide between those two and would like your input.

pingpongboss replied:
For a while now, this thought has been creeping in my mind. Maybe having multiple SynchronizationMechanism that the client provides is not the right pattern. 

An alternative pattern is for Remixer to use a single implementation of SynchronizationMechanism, but allow that SynchronizationMechanism to be configured by the client. Eg:

```
class Remixer {
  private final SynchronizationMechanism sync = new RemixerSynchronizationMechanism();

  public SynchronizationMechanism getSynchronizationMechanism() {
    return this.sync;
  }
}
```

```
class ClientApplication extends Application {
  void onCreate() {
    super.onCreate();

    Remixer.getInstance().getSynchronizationMechanism().sharedPreferencesSupport = true;
    Remixer.getInstance().getSynchronizationMechanism().firebaseSupport = false;
  }
}
```

This alternative proposal would get rid of a lot of the "icky feelings" I have about the current pattern (those feelings are hard to organize into words).

A huge advantage to this is exactly what you said in your message. It would allow the SynchronizationMechanism to own more of this logic that I'd like to see pulled out of Remixer. Right now there's a little bit of syncing logic in Remixer and a little bit in SynchronizationMechanism. I'd like to see all of it go to SynchronizationMechanism.

This would include:
- When a new variable is added, the SynchronizationMechanism checks all others to make sure the new variable is legal.
- When a variable value changes, the SynchronizationMechanism updates all other equivalent variables.

miguelandres replied:
I think separate implementations are still necessary, and I would avoid duplicating code by making this localvaluesyncing class the base class for all of them. 

The reason for this is that local-only and cloud mode kinda have different rules RE: source of truth and stuff like that, that would make the code really hard to read in a single configurable object.

miguelandres replied:
But still move the value syncing code on modifications to the synchronization mechanism (done already)

pingpongboss reviewed (commented):
Here is a partial review.  It's a little confusing at the moment what Remixer is in charge of and what the synchronizationMechanism is in charge of.

Commit:
Improve Remixer Documenttion and guard against resetting the SynchronizationMechanism

Commit:
Merge branch 'develop' into move_sync_logic

Commit:
Fix reference to RemixerActivityLifecycleCallbacks in the README

pingpongboss commented on README.md:
```diff
@@ -55,13 +57,165 @@ And in your modules, apply the `android-apt` plugin and add the remixer dependen
 Notice the dependency on `remixer_annotation` is a `provided` clause instead of `compile`, this is on purpose as this is not a regular dependency but a compiler plugin.
 
+### Global remixer set up
+If you have not subclassed the application class it is recommended you do it since this is a one-time global initialization.
```
"it is recommended" or it is required to use remixer?

pingpongboss commented on README.md:
```diff
@@ -55,13 +57,165 @@ And in your modules, apply the `android-apt` plugin and add the remixer dependen
+### Global remixer set up
+If you have not subclassed the application class it is recommended you do it since this is a one-time global initialization.
+
+In your application class you need to add RemixerActivityLifecycleCallbacks as an ActivityLifecycleCallbacks instance, so Remixer knows to remove old variables and triggers when activities are destroyed, to avoid leaks.
```
my advice is to reduce the explanation from the README (clients just want to know how to get things to work) and add the explanation to `RemixerActivityLifecycleCallbacks`'s javadoc (that's where clients would look if they were curious)

pingpongboss replied:
Or, have a "how to get started section at the top of the README, with the quickest instructions on how to integrate remixer". Then have a more detailed section at the bottom of the README.

pingpongboss commented on README.md:
```diff
@@ -55,13 +57,165 @@ And in your modules, apply the `android-apt` plugin and add the remixer dependen
+  - A variation of `remixerFragment.attachToGesture(this, Direction.UP, 3);`, this example ties showing the Remixer Fragment on a 3-finger swipe up.
+  - `remixerFragment.attachToButton(this, someButtonObject);`, this makes the OnClickListener for a button open the Remixer fragment.
+
+Your `Activity.onCreate` may look like this:
```
same advice for this section. Although you may want to keep _some_ explanation, most of this can be replaced with "Learn more about the APIs defined in the library by reading our [technical documentation](https://jitpack.io/com/github/material-foundation/material-remixer-android/0.5.0/javadoc/)" (you can add the link once you have javadoc generated)

pingpongboss commented on README.md:
```diff
@@ -55,13 +57,165 @@ And in your modules, apply the `android-apt` plugin and add the remixer dependen
+
+In order to define variables you only need to write methods that take one argument of the correct type and annotate them. The methods contain your logic to handle changes to these variables (update the UI accordingly). You can rest assured those methods will run in the main UI thread.
+
+There are a few very simple examples here, but you should look at the [example](https://github.com/material-foundation/material-remixer-android/blob/develop/remixer_example/src/main/java/com/google/android/apps/remixer/MainActivity.java) [activities](https://github.com/material-foundation/material-remixer-android/blob/develop/remixer_example/src/main/java/com/google/android/apps/remixer/BoxActivi ...
```
I wouldn't link [example] [activities] in that way for clarity reasons. Maybe just link both words to `https://github.com/material-foundation/material-remixer-android/blob/develop/remixer_example/src/main/java/com/google/android/apps/remixer`

pingpongboss commented on README.md:
```diff
@@ -55,13 +57,165 @@ And in your modules, apply the `android-apt` plugin and add the remixer dependen
+
+There are a few very simple examples here, but you should look at the [example](https://github.com/material-foundation/material-remixer-android/blob/develop/remixer_example/src/main/java/com/google/android/apps/remixer/MainActivity.java) [activities](https://github.com/material-foundation/material-remixer-android/blob/develop/remixer_example/src/main/java/com/google/android/apps/remixer/BoxActivi ...
+
+Once you add your annotated methods and build you should be able to invoke remixer (by doing a 3 finger swipe or clicking a button, depending on how you configured it in the section above), and tweak the variables or trigger the events guarded by the trigger.
```
The "getting started" section is missing an easily copyable code snippet.

pingpongboss commented on README.md:
```diff
@@ -55,13 +57,165 @@ And in your modules, apply the `android-apt` plugin and add the remixer dependen
+Once you add your annotated methods and build you should be able to invoke remixer (by doing a 3 finger swipe or clicking a button, depending on how you configured it in the section above), and tweak the variables or trigger the events guarded by the trigger.
+
+##### Range variables
+__Note:__ for the time being they only support Integers, not floats or doubles.
```
Ah good to hear. Insert link to issue in README.

pingpongboss commented on remixer_core/src/main/java/com/google/android/libraries/remixer/Remixer.java:
```diff
@@ -16,13 +16,25 @@
+ *
+ * <p>If you do not set a SynchronizationMechanism, remixer will not try to sync values across
+ * different contexts (activities) or check for consistency among those. It is recommended you
+ * always use one (even if it is just the LocalValueSyncing one) so you catch consistency errors.
```
So this makes me wonder. Why doesn't the Remixer object just always initialize 

```
this. synchronizationMechanism = new LocalValueSyncing();
```

This way, clients don't have to be bothered to initialize in the 99% case ("It is recommended you always use one") and you get consistency checks for free. If clients do want to set their own synchronizationMechanism, they can easily do so in Application.onCreate()

miguelandres replied:
This is a good idea, I think I'm going to do this instead then :) Thanks.

miguelandres replied:
Ugh, no, can't unless we move the whole remixer_storage into remixer_core, because LocalValueSyncing is in remixer_storage which depends on remixer_core, can't have a circular dependency.

We have two options:

# Move the current contents of remixer_storage into remixer_core

-Makes it possible to have a default
-Makes it possible to keep some of those private methods package-local
-Kinda pollutes remixer_core with serialization logic.
-Remixer_storage/sync will still be necessary because firebase cloud sdk has to be used from an android_library, so some sync logic would be in remixer_core and some in remixer_storage/sync.

# Keep it as is

-storage and sync is all centralized in a separate module, does not pollute remixer_core
-impossible to have a default.
-need to keep some of those methods public (triggerWithoutNotifiyingOthersWithTheSameKey, setValueWithoutNotifyingOthersWithTheSameKey) :/

# Conclusion

I kinda prefer moving this sync logic to remixer_core, even if it feels kinda out of place when the things that depend on android SDK are put in a separate project.

What do you think?

pingpongboss replied:
What's the current status on this? Have you re-evaluated if `LocalValueSyncing` should be provided by default?

miguelandres replied:
Sorry, forgot to update this comment. Yes, it's provided by default.

pingpongboss commented on remixer_core/src/main/java/com/google/android/libraries/remixer/Remixer.java:
```diff
@@ -16,13 +16,25 @@
+ *
+ * <p>For value and data syncing/persistence (both locally across activities, and persisting/saving
+ * to cloud backends, etc) and keeping a global state, the Remixer object relies on a {@link
+ * SynchronizationMechanism} set at initialization time (usually in your Application.onCreate()).
```
I just want to bring this advice to your attention. Is it possible that Application.onCreate() can be called more than once? Then remixer would crash since it is guarding in `setSynchronizationMechanism()`

miguelandres replied:
It would only be called once.

miguelandres replied:
If it's called again the application has been killed and restarted, that's guaranteed.

Commit:
Merge branch 'develop' into move_sync_logic

Commit:
Merge branch 'develop' into move_sync_logic

Commit:
Move the notifying other variables of changes and other triggers out of remixer

This makes more sense in the SynchronizationMechanism.

Commit:
Fix lint issues

Commit:
Merge branch 'develop' into move_sync_logic

Commit:
Remove creating mapping logic from getItemsWithKey and getItemsWithContext.

Split this into a separate method, getOrCreateItemList.

Commit:
Remove creating mapping logic from getItemsWithKey and getItemsWithContext (renamed).

Move this logic to a helper private method.

Commit:
Merge branch 'add_context_map' into move_sync_logic

Commit:
Merge branch 'develop' into move_sync_logic

Commit:
Add a type-safe data type to RemixerItem.

This is still incomplete. Annotation-processing is not working because it hasn’t been adapted to use it. I am planning to move to use the Builders first so it’s easier to adapt.

Commit:
Make annotation processing use the Builders for every annotation.

This large change cleans up a lot of the annotation processing code, making a lot of it more reusable. It means all of the common fields for a Remixer Item are initialized by `MethodAnnotation` instead of being initialized all over the place. It improves readability and maintainability of this component, and lets us more easily adapt to changes in the API.

In order to do this I had to add a builder for Trigger that lacked one and make all builders inherit from a `RemixerItem.Builder` class.

Commit:
Merge branch 'develop' into add_types

Commit:
Merge branch 'develop' into add_types

Commit:
Initial merge of builders in annotations, missing the data type in tests

Commit:
Move all tests in remixer_core and remixer_ui to use Builders

Commit:
Merge branch 'move_all_tests_to_use_builders' into add_types

Commit:
Merge branch 'develop' into add_types

Commit:
Merge remote-tracking branch 'origin/develop' into add_types

Commit:
Merge branch 'develop' into move_sync_logic

Commit:
Merge branch 'add_types' into move_sync_logic

Commit:
Move all of the serialization logic into `remixer_core`

That allows us to use it directly in `remixer_core` to improve variable value synchronization and build better foundation for cloud sync.

Commit:
Merge branch 'develop' into move_serialization

Commit:
Merge branch 'move_serialization' into move_sync_logic

Commit:
checkstyle fixes

miguelandres commented:
This has now been synced to the changes up to #89, depends on #89, but works now, this addresses most of the comments already here, you should take another look since it's different enough.

Commit:
Merge remote-tracking branch 'origin/develop' into move_sync_logic

# Conflicts:
#	README.md
#	remixer_core/src/main/java/com/google/android/libraries/remixer/DataType.java
#	remixer_core/src/main/java/com/google/android/libraries/remixer/Remixer.java
#	remixer_core/src/main/java/com/google/android/libraries/remixer/serialization/SerializableRemixerContents.java
#	remixer_core/src/main/java/com/google/android/libraries/remixer/serialization/StoredVariable.java
#	remixer_core/src/test/java/com/google/android/libraries/remixer/serialization/BooleanVariableSerializationTest.java
#	remixer_core/src/test/java/com/google/android/libraries/remixer/serialization/ColorVariableSerializationTest.java
#	remixer_core/src/test/java/com/google/android/libraries/remixer/serialization/IntegerVariableSerializationTest.java
#	remixer_core/src/test/java/com/google/android/libraries/remixer/serialization/SerializableRemixerContentsSerializationTest.java
#	remixer_core/src/test/java/com/google/android/libraries/remixer/serialization/StringVariableSerializationTest.java

Commit:
Fix bug with different titles accross contexts

Commit:
Fix comment on default SynchronizationMechanism

pingpongboss commented on remixer_core/src/main/java/com/google/android/libraries/remixer/sync/LocalValueSyncing.java:
```diff
@@ -0,0 +1,75 @@
+public class LocalValueSyncing implements SynchronizationMechanism {
+
+  SerializableRemixerContents serializableRemixerContents = new SerializableRemixerContents();
+  Remixer remixer;
```
can these guys be private?

miguelandres replied:
Done.

pingpongboss commented on remixer_core/src/test/java/com/google/android/libraries/remixer/InitializationHelper.java:
```diff
@@ -0,0 +1,36 @@
+ */
+public class InitializationHelper {
+
+  public static void init() {
```
add documentation to this. You expect clients to call this in their Application right?

pingpongboss replied:
I wonder if you can "catch" it if clients forget. If Remixer has 0 registered data types, it should throw an exception?

miguelandres replied:
This is a test helper, the real version is in RemixerInitialization and it does a few more things (initialize default UI mappings and so on).

You already filed a bug to catch uninitialized remixer instances early, I'm thinking of some way to do this, but will follow up on this on that bug.

pingpongboss reviewed (approved):
Looks awesome! Mostly small comments. Big things can be done in follow up if necessary.

Commit:
Fix visibility and comments
